### PR TITLE
Supply Euler angles from aux-variables for CP

### PR DESCRIPTION
**Pull request by dewenyushu**
Add an option in `ComputeElasticityTensorCP` for getting Euler angle values from `AuxVariables`. This would be useful for coupling crystal plasticity with phase-field simulations.

### CODE REVIEW

File: modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C [hunk 1]
```diff
@@ -31,7 +33,9 @@ ComputeElasticityTensorCP::ComputeElasticityTensorCP(const InputParameters & par
                                : nullptr),
     _Euler_angles_mat_prop(declareProperty<RealVectorValue>("Euler_angles")),
     _crysrot(declareProperty<RankTwoTensor>(_base_name + "crysrot")),
-    _R(_Euler_angles)
+    _R(_Euler_angles),
+    _n_euler_angle_vars(
+        isCoupled("euler_angle_variables") ? coupledComponents("euler_angle_variables") : 0)
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C:38]: is coupled components not simply 0 if isCoupled is false?

File: modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C [hunk 2]
```diff
@@ -56,6 +60,22 @@ ComputeElasticityTensorCP::ComputeElasticityTensorCP(const InputParameters & par
...
+
+  if (isCoupled("euler_angle_variables") && _n_euler_angle_vars != 3)
+    paramError("euler_angle_variables", "The Euler angles should have three components.");
+
+  if (isCoupled("euler_angle_variables"))
+  {
+    for (unsigned int i = 0; i < _n_euler_angle_vars; i++)
+      _euler_angle_vars.push_back(
+          dynamic_cast<MooseVariable *>(getVar("euler_angle_variables", i)));
+  }
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C:78]: we have `coupledValues` to do this with one call

Comment (dschwen) [reply modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C:78]: https://github.com/idaholab/moose/blob/8465925da35e576d75247f34c562ed4889214cc0/framework/include/interfaces/Coupleable.h#L205

File: modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C [hunk 3]
```diff
@@ -67,6 +87,12 @@ ComputeElasticityTensorCP::assignEulerAngles()
     _Euler_angles_mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
     _Euler_angles_mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);
   }
+  else if (_n_euler_angle_vars)
+  {
+    _Euler_angles_mat_prop[_qp](0) = _euler_angle_vars[0]->sln()[_qp];
+    _Euler_angles_mat_prop[_qp](1) = _euler_angle_vars[1]->sln()[_qp];
+    _Euler_angles_mat_prop[_qp](2) = _euler_angle_vars[2]->sln()[_qp];
```
Comment (dschwen) [inline modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C:94]: why use `sln()` here and not just a coupled value?

**Review by dschwen — COMMENTED** [review]
Questions about coupling

File: modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C (hunk 1 above)
Comment (dewenyushu) [reply modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C]: That's right! - addressed.

File: modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C (hunk 2 above)
Comment (dewenyushu) [reply modules/solid_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C]: Thanks for the suggestions - yes the coupledValues is good for this case. Addressed

**Review by dschwen — APPROVED** [review]

**Comment by dewenyushu** [comment]
@dschwen this should be ready!

File: modules/solid_mechanics/include/materials/crystal_plasticity/ComputeElasticityTensorCP.h [hunk 4]
```diff
@@ -52,4 +52,8 @@ class ComputeElasticityTensorCP : public ComputeElasticityTensor
 
   /// flag for user-defined rotation matrix, supplied in input file
   bool _user_provided_rotation_matrix;
+
+  // The coupled Euler angles component variables
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/crystal_plasticity/ComputeElasticityTensorCP.h:56]:
```suggestion
  /// The coupled Euler angles component variables
```

**Pull request merged**